File: src/discovery/linkedin_api.py
import hashlib
import json
import re
from datetime import datetime
from typing import List, Dict, Optional

from apify_client import ApifyClient
from src import config
from src.utils.logger import log, error

client = ApifyClient(config.APIFY_TOKEN)

# Actor ID for the LinkedIn Jobs Scraper by automation-lab
LINKEDIN_ACTOR_ID = "automation-lab/linkedin-jobs-scraper"


def stable_job_id(prefix: str, value: str) -> str:
    digest = hashlib.sha1(value.encode("utf-8")).hexdigest()[:16]
    return f"{prefix}-{digest}"


def normalize_linkedin_job(item: Dict) -> Dict:
    """
    Normalize a raw LinkedIn job item from the Apify dataset
    into the shared Job-Bot schema.
    """
    title = item.get("title", "")
    location = item.get("location", "")
    
    is_remote = (
        bool(re.search(r"remote", title, re.IGNORECASE)) or 
        bool(re.search(r"remote", location, re.IGNORECASE)) or 
        item.get("workplaceType") == "Remote"
    )

    return {
        "source": "LinkedIn",
        "job_title": item.get("title", "Data Engineering"),
        "company": item.get("companyName", "Unknown"),
        "company_url": item.get("companyLinkedinUrl"),
        "company_logo": item.get("companyLogo"),
        "location": location or "United States",
        "remote": is_remote,
        "posted_date": item.get("postedAt"),
        "scraped_at": item.get("scrapedAt", datetime.utcnow().isoformat()),
        "url": item.get("url"),
        "apply_url": item.get("applyUrl") or item.get("url"),
        "description": item.get("descriptionText", ""),
        "description_html": item.get("descriptionHtml", ""),
        "requirements": None,
        "salary": item.get("salary"),
        "seniority_level": item.get("seniorityLevel"),
        "employment_type": item.get("employmentType"),
        "job_function": item.get("jobFunction"),
        "industries": item.get("industries"),
        "applicants_count": item.get("applicantsCount"),
        "benefits": item.get("benefits"),
        "id": stable_job_id("linkedin", str(item.get("id") or item.get("url") or item.get("title") or "")),
        "raw_payload": item,
    }

def fetch_linkedin_jobs_via_apify_api(options: Optional[Dict] = None) -> List[Dict]:
    """
    Fetch LinkedIn Data Engineering jobs via the Apify API.
    Uses the automation-lab/linkedin-jobs-scraper actor.
    """
    if options is None:
        options = {}

    try:
        run_input = {
            "searchQuery": options.get("searchQuery", config.JOB_KEYWORDS),
            "location": options.get("location", config.JOB_LOCATION),
            "maxJobs": options.get("maxJobs", 999),
            "workplaceType": options.get("workplaceType", "2"),  # Remote
            "datePosted": options.get("datePosted", "r86400"),  # Last 24 hours
            "sortBy": options.get("sortBy", "R"),              # Most recent
            "experienceLevel": options.get("experienceLevel", "all"),
            "jobType": options.get("jobType", "all"),
            "scrapeJobDetails": options.get("scrapeJobDetails", True),
            "maxRequestRetries": config.APIFY_MAX_RETRIES,
        }

        log(f"Starting LinkedIn Apify actor ({LINKEDIN_ACTOR_ID}) with input:")
        log(json.dumps(run_input, indent=2))

        run = client.actor(LINKEDIN_ACTOR_ID).call(
            run_input=run_input,
            timeout_secs=config.APIFY_TIMEOUT_SECONDS,
            memory_mbytes=256,
        )

        jobs = []
        if run and run.get("defaultDatasetId"):
            log(f"Actor run completed. Dataset ID: {run['defaultDatasetId']}")
            dataset_items = client.dataset(run["defaultDatasetId"]).list_items().items
            
            for item in dataset_items:
                jobs.append(normalize_linkedin_job(item))
        else:
            error("Actor run completed but no dataset was returned.")

        log(f"LinkedIn API fetch completed: {len(jobs)} jobs found")
        return jobs

    except Exception as err:
        error(f"LinkedIn API fetch failed: {str(err)}")
        return []

def fetch_linkedin_jobs_from_dataset(dataset_id: str) -> List[Dict]:
    """
    Fetch jobs from a previous LinkedIn actor run's dataset.
    """
    try:
        log(f"Fetching jobs from existing dataset: {dataset_id}")
        dataset_items = client.dataset(dataset_id).list_items().items
        jobs = [normalize_linkedin_job(item) for item in dataset_items]
        
        log(f"Fetched {len(jobs)} jobs from dataset {dataset_id}")
        return jobs
    except Exception as err:
        error(f"Failed to fetch from dataset {dataset_id}: {str(err)}")
        return []

File: src/discovery/linkedin-api.js
import { ApifyClient } from "apify-client";
import config from "../config.js";
import { log, error } from "../utils/logger.js";

const client = new ApifyClient({ token: config.apifyToken });

// Actor ID for the LinkedIn Jobs Scraper by automation-lab
const LINKEDIN_ACTOR_ID = "automation-lab/linkedin-jobs-scraper";

/**
 * Normalize a raw LinkedIn job item from the Apify dataset
 * into the shared Job-Bot schema.
 */
function normalizeLinkedInJob(item) {
  return {
    source: "LinkedIn",
    job_title: item.title || "Data Engineering",
    company: item.companyName || "Unknown",
    company_url: item.companyLinkedinUrl || null,
    company_logo: item.companyLogo || null,
    location: item.location || "United States",
    remote: /remote/i.test(item.title || "") || /remote/i.test(item.location || "") || item.workplaceType === "Remote",
    posted_date: item.postedAt || null,
    scraped_at: item.scrapedAt || new Date().toISOString(),
    url: item.url || null,
    apply_url: item.applyUrl || item.url || null,
    description: item.descriptionText || "",
    description_html: item.descriptionHtml || "",
    requirements: null,
    salary: item.salary || null,
    seniority_level: item.seniorityLevel || null,
    employment_type: item.employmentType || null,
    job_function: item.jobFunction || null,
    industries: item.industries || null,
    applicants_count: item.applicantsCount || null,
    benefits: item.benefits || null,
    id: `linkedin-${item.id || encodeURIComponent(item.url || "")}`,
    raw_payload: item,
  };
}

/**
 * Fetch LinkedIn Data Engineering jobs via the Apify API.
 * Uses the automation-lab/linkedin-jobs-scraper actor.
 *
 * @param {Object} options - Override default search parameters
 * @param {string} options.searchQuery - Search keywords (default: from config)
 * @param {string} options.location - Location filter (default: from config)
 * @param {number} options.maxJobs - Max jobs to scrape (default: 999)
 * @param {string} options.workplaceType - "2" = remote, "1" = on-site, "3" = hybrid
 * @param {string} options.datePosted - "r86400" = last 24h, "r604800" = last week
 * @param {boolean} options.scrapeJobDetails - Whether to scrape full descriptions
 * @returns {Promise<Array>} Normalized job objects
 */
export async function fetchLinkedInJobsViaApifyAPI(options = {}) {
  try {
    const input = {
      searchQuery: options.searchQuery || config.jobKeywords,
      location: options.location || config.jobLocation,
      maxJobs: options.maxJobs || 999,
      workplaceType: options.workplaceType || "2",           // Remote
      datePosted: options.datePosted || "r86400",             // Last 24 hours
      sortBy: options.sortBy || "R",                          // Most recent
      experienceLevel: options.experienceLevel || "all",
      jobType: options.jobType || "all",
      scrapeJobDetails: options.scrapeJobDetails !== undefined ? options.scrapeJobDetails : true,
      maxRequestRetries: config.apify.maxRetries,
    };

    log(`Starting LinkedIn Apify actor (${LINKEDIN_ACTOR_ID}) with input:`);
    log(JSON.stringify(input, null, 2));

    const run = await client.actor(LINKEDIN_ACTOR_ID).call(input, {
      timeoutSecs: config.apify.timeoutSeconds,
      memoryMbytes: 256,
    });

    const jobs = [];

    if (run && run.defaultDatasetId) {
      log(`Actor run completed. Dataset ID: ${run.defaultDatasetId}`);
      const dataset = await client.dataset(run.defaultDatasetId).listItems();

      if (dataset && dataset.items) {
        for (const item of dataset.items) {
          jobs.push(normalizeLinkedInJob(item));
        }
      }
    } else {
      error("Actor run completed but no dataset was returned.");
    }

    log(`LinkedIn API fetch completed: ${jobs.length} jobs found`);
    return jobs;
  } catch (err) {
    error(`LinkedIn API fetch failed: ${err.message}`);
    return [];
  }
}

/**
 * Fetch jobs from a previous LinkedIn actor run's dataset.
 * Useful for re-processing data without re-running the actor.
 *
 * @param {string} datasetId - The Apify dataset ID from a previous run
 * @returns {Promise<Array>} Normalized job objects
 */
export async function fetchLinkedInJobsFromDataset(datasetId) {
  try {
    log(`Fetching jobs from existing dataset: ${datasetId}`);
    const dataset = await client.dataset(datasetId).listItems();
    const jobs = [];

    if (dataset && dataset.items) {
      for (const item of dataset.items) {
        jobs.push(normalizeLinkedInJob(item));
      }
    }

    log(`Fetched ${jobs.length} jobs from dataset ${datasetId}`);
    return jobs;
  } catch (err) {
    error(`Failed to fetch from dataset ${datasetId}: ${err.message}`);
    return [];
  }
}
